src/node.h

```diff
@@ -382,13 +382,14 @@ NODE_EXTERN v8::Isolate* NewIsolate(
 
 // Creates a new context with Node.js-specific tweaks.
 NODE_EXTERN v8::Local<v8::Context> NewContext(
-    v8::Isolate* isolate,
+    IsolateData* isolate_data,
```

Comment:
This seems to be a breaking change for the embedders. I wonder if there is a way to get around it. For example, can we store the IsolateData* in the Isolate via `Isolate::SetData()` in `CreateIsolateData` and use `Isolate::GetData()` to retrieve it? Then `NewContext` would be able to recognize any Isolate that goes through `CreateIsolateData` before.

It does expose the issue about how we can run `NewContext` on an isolate that has never go through `CreateIsolateData` before, or specifically, how do you generate a Node.js-aware context with an Isolate that's not even Node.js aware. I think for embedders' use case it probably makes more sense to just run `CreateIsolateData` in `NewIsolate` - which means you can only set up a Node.js-aware context with Node.js-aware Isolate, which makes sense to me. We can handle the existing public `CreateIsolateData` gracefully by returning if the IsolateData is already created. Also we can probably just deprecate `CreateIsolateData` too, internally we always run it before `NewContext`, but the existing test runs it after `NewContext`, which IMO is a rather odd use case that can be dropped.

Reply:
Another way to avoid this that would involve much fewer changes is to simply remove the symbols from IsolateData, and use `v8::Symbol::For` as well as the JS counterpart to retrieve these symbols. If we really want to hide them from users (I don't see why though, ideally don't we even want to expose them for users to implement their own transferrables?), we can store them in the context with keys generated with `v8::Private::ForApi` too.

Reply:
@joyeecheung I like the second option more because it takes us one step closer to implementing https://github.com/nodejs/node/issues/37080. Exposed the global symbols in https://github.com/nodejs/node/pull/41148/commits/1fc8927714e819d67c16f8474c6a02086da10f42 and reverted the embedder breaking changes in https://github.com/nodejs/node/pull/41148/commits/f7b4f9bde9a699809473b85bc7ff3a2c31fb700d, PTAL.